File: app/src/main/java/com/vikingsen/cheesedemo/model/data/price/PriceRemoteDataSource.java
package com.vikingsen.cheesedemo.model.data.price;


import com.vikingsen.cheesedemo.model.webservice.CheeseService;
import com.vikingsen.cheesedemo.model.webservice.dto.PriceDto;
import com.vikingsen.cheesedemo.util.NetworkDisconnectedException;
import com.vikingsen.cheesedemo.util.NetworkUtil;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Single;
import retrofit2.Response;
import timber.log.Timber;

@Singleton
class PriceRemoteDataSource {

    private final CheeseService cheeseService;
    private final NetworkUtil networkUtil;

    @Inject
    PriceRemoteDataSource(CheeseService cheeseService,
                          NetworkUtil networkUtil) {
        this.cheeseService = cheeseService;
        this.networkUtil = networkUtil;
    }

    Single<PriceDto> getPrice(long cheeseId) {
        return Single.create(emitter -> {
            try {
                if (networkUtil.isConnected()) {
                    Response<PriceDto> response = cheeseService.getPrice(cheeseId).execute();
                    if (response.isSuccessful()) {
                        emitter.onSuccess(response.body());
                    } else {
                        Timber.e("Failed to load price for cheese %d (%s) : %s", cheeseId, response.code(), response.message());
                        emitter.onError(new PriceFailureException());
                    }
                } else {
                    Timber.w("Network not connected");
                    emitter.onError(new NetworkDisconnectedException());
                }
            } catch (Exception e) {
                Timber.e(e, "Exception fetching price for cheese %d", cheeseId);
                emitter.onError(new PriceFailureException(e));
            }
        });
    }
}

File: app/src/main/java/com/vikingsen/cheesedemo/model/data/price/PriceRemoteDataSource.kt
package com.vikingsen.cheesedemo.model.data.price


import android.arch.lifecycle.LiveData
import com.vikingsen.cheesedemo.model.webservice.ApiResponse
import com.vikingsen.cheesedemo.model.webservice.CheeseService
import com.vikingsen.cheesedemo.model.webservice.dto.PriceDto
import com.vikingsen.cheesedemo.util.CoroutineContextProvider
import com.vikingsen.cheesedemo.util.NetworkDisconnectedException
import com.vikingsen.cheesedemo.util.NetworkUtil
import kotlinx.coroutines.experimental.delay
import kotlinx.coroutines.experimental.launch
import timber.log.Timber
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PriceRemoteDataSource
@Inject constructor(
        private val cheeseService: CheeseService,
        private val networkUtil: NetworkUtil,
        private val ccp: CoroutineContextProvider
) {

    fun getPrice(cheeseId: Long): LiveData<ApiResponse<PriceDto>> {
        return object : LiveData<ApiResponse<PriceDto>>() {
            private val executed = AtomicBoolean(false)

            override fun onActive() {
                if (executed.compareAndSet(false, true)) {
                    execute()
                }
            }

            private fun execute() {
                launch(ccp.commonPool) {
                    postValue(null)
                    // Fake Network Time
                    delay(2, TimeUnit.SECONDS)
                    try {
                        if (networkUtil.isConnected) {
                            postValue(ApiResponse(cheeseService.getPrice(cheeseId).execute()))
                            return@launch
                        }
                    } catch (e: Exception) {
                        Timber.e(e, "Exception fetching cheese")
                    }
                    postValue(ApiResponse(NetworkDisconnectedException()))
                }
            }
        }
    }
}
